pkg/operation/botanist/kubeapiserver.go

```diff
@@ -156,6 +191,72 @@ func (b *Botanist) computeKubeAPIServerAdmissionPlugins(defaultPlugins, configur
 	return admissionPlugins
 }
 
+func (b *Botanist) ensureAdmissionPluginConfig(plugins []gardencorev1beta1.AdmissionPlugin, kubernetesVersion string) ([]gardencorev1beta1.AdmissionPlugin, error) {
+	var (
+		index                     int
+		allowPrivilegedContainers = pointer.BoolDeref(b.Shoot.GetInfo().Spec.Kubernetes.AllowPrivilegedContainers, false)
+		pspDisabled               = b.Shoot.PSPDisabled
+	)
+
+	for i, plugin := range plugins {
+		if plugin.Name == "PodSecurity" {
+			index = i
+		}
+	}
+
+	// if a config is set in the shoot spec
+	if plugins[index].Config != nil {
+		config, err := runtime.Decode(codec, plugins[index].Config.Raw)
+		if err != nil {
+			return plugins, err
+		}
+		admConfig, ok := config.(*admissionapiv1beta1.PodSecurityConfiguration)
+		if !ok {
+			return plugins, fmt.Errorf("PodSecurity admission config is not of type PodSecurityConfiguration")
+		}
+
+		// exempt kube-system namespace
+		if !slices.Contains(admConfig.Exemptions.Namespaces, metav1.NamespaceSystem) {
+			admConfig.Exemptions.Namespaces = append(admConfig.Exemptions.Namespaces, metav1.NamespaceSystem)
+		}
+		// if allowPrivilegedContainers is false, enforce restricted level
+		if versionutils.ConstraintK8sGreaterEqual123.Check(semver.MustParse(kubernetesVersion)) &&
+			pspDisabled &&
+			!allowPrivilegedContainers {
+			admConfig.Defaults.Enforce = "restricted"
+			admConfig.Defaults.EnforceVersion = "latest"
+		}
+
+		configData, err := runtime.Encode(codec, admConfig)
+		if err != nil {
+			return plugins, err
+		}
+		plugins[index].Config = &runtime.RawExtension{Raw: configData}
+	} else {
+		// if no config is set and allowPrivilegedContainers is false, add default config
+		if versionutils.ConstraintK8sGreaterEqual123.Check(semver.MustParse(kubernetesVersion)) &&
+			pspDisabled &&
+			!allowPrivilegedContainers {
+			defaultConfig := &admissionapiv1beta1.PodSecurityConfiguration{
+				Defaults: admissionapiv1beta1.PodSecurityDefaults{
+					Enforce:        "restricted",
+					EnforceVersion: "latest",
+				},
+				Exemptions: admissionapiv1beta1.PodSecurityExemptions{
+					Namespaces: []string{metav1.NamespaceSystem},
+				},
+			}
+			defaultConfigData, err := runtime.Encode(codec, defaultConfig)
+			if err != nil {
+				return plugins, err
+			}
+			plugins[index].Config = &runtime.RawExtension{Raw: defaultConfigData}
```

Comment:
This is a lot of duplication with the above branch. Can you harmonize this and DRY?